File: cpp/src/Freeze/Application.cpp
// **********************************************************************
//
// Copyright (c) 2001
// Mutable Realms, Inc.
// Huntsville, AL, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <Freeze/Application.h>

using namespace std;
using namespace Ice;
using namespace Freeze;

Freeze::Application::Application(const string& dbEnvName) :
    _dbEnvName(dbEnvName)
{
}

Freeze::Application::~Application()
{
}

int
Freeze::Application::run(int argc, char* argv[])
{
    ignoreInterrupt();
    
    int status;
    DBEnvironmentPtr dbEnv;
    
    try
    {
	dbEnv = Freeze::initialize(communicator(), _dbEnvName);
	status = runFreeze(argc, argv, dbEnv);
    }
    catch(const DBException& ex)
    {
	cerr << appName() << ": " << ex << ": " << ex.message << endl;
	status = EXIT_FAILURE;
    }
    // TODO: How does this compile?
    catch(const Exception& ex)
    {
	cerr << appName() << ": " << ex << endl;
	status = EXIT_FAILURE;
    }
    catch(...)
    {
	cerr << appName() << ": unknown exception" << endl;
	status = EXIT_FAILURE;
    }

    if(dbEnv)
    {
	try
	{
	    dbEnv->close();
	}
	catch(const DBException& ex)
	{
	    cerr << appName() << ": " << ex << ": " << ex.message << endl;
	    status = EXIT_FAILURE;
	}
	catch(const Exception& ex)
	{
	    cerr << appName() << ": " << ex << endl;
	    status = EXIT_FAILURE;
	}
	catch(...)
	{
	    cerr << appName() << ": unknown exception" << endl;
	    status = EXIT_FAILURE;
	}
	dbEnv = 0;
    }

    defaultInterrupt();
    return status;
}

File: java/src/Freeze/Application.java
// **********************************************************************
//
// Copyright (c) 2002
// ZeroC, Inc.
// Billerica, MA, USA
//
// All Rights Reserved.
//
// Ice is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License version 2 as published by
// the Free Software Foundation.
//
// **********************************************************************

package Freeze;

public abstract class Application extends Ice.Application
{
    public
    Application(String dbEnvName)
    {
	_dbEnvName = dbEnvName;
    }

    public int
    run(String[] args)
    {
	int status;
	DBEnvironment dbEnv = null;

	try
	{
	    dbEnv = Freeze.Util.initialize(communicator(), _dbEnvName);
	    status = runFreeze(args, dbEnv);
	}
	catch(DBException ex)
	{
	    System.err.println(appName() + ": " + ex + ": " + ex.message);
	    status = 1;
	}
	catch(Ice.LocalException ex)
	{
	    System.err.println(appName() + ": " + ex);
	    ex.printStackTrace();
	    status = 1;
	}
	catch(Exception ex)
	{
	    System.err.println(appName() + ": unknown exception");
	    ex.printStackTrace();
	    status = 1;
	}
	
	if(dbEnv != null)
	{
	    try
	    {
		dbEnv.close();
	    }
	    catch(DBException ex)
	    {
		System.err.println(appName() + ": " + ex + ": " + ex.message);
		ex.printStackTrace();
		status = 1;
	    }
	    catch(Exception ex)
	    {
		System.err.println(appName() + ": unknown exception");
		ex.printStackTrace();
		status = 1;
	    }
	    dbEnv = null;
	}
	
	return status;
    }

    public abstract int
    runFreeze(String[] args, DBEnvironment dbEnv);

    private String _dbEnvName;
}
